csrc/multidevice/utils.cpp

```diff
@@ -234,41 +223,62 @@ void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
   }
   if (!tvs.empty()) {
     scheduler_utils::parallelizeAllLike(
-        ref,
-        tvs,
-        {ParallelType::DIDx, ParallelType::Serial},
-        /*propagate_padding=*/false,
-        /*skip_reductions=*/true);
+        ref, tvs, {ParallelType::DIDx, ParallelType::Serial});
+  }
+}
+
+namespace {
+void propagateShardings(
+    const std::vector<TensorView*>& from_tvs,
+    const std::unordered_set<TensorView*>& boundary_tvs,
+    TensorView* ref) {
+  auto tvs_between = scheduler_utils::getAllTvsFrom(from_tvs, boundary_tvs);
+  // std::vector<TensorView*> tvs;
+  // std::copy_if(tvs_between.begin(), tvs_between.end(),
+  // std::back_inserter(tvs), [](TensorView* tv) {return !isDIDReduction(tv);});
+  shardAllLike(ref, {tvs_between.begin(), tvs_between.end()});
+
+  // Remove DID parallelizations on reduction axes.
+  for (auto tv : tvs_between) {
+    for (auto id : tv->getLogicalDomain()) {
```

Comment:
```suggestion
    for (IterDomain* id : tv->domain()->getAllIDs()) {
```
so it's still correct when we split loop instead.